tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +215,130 @@ func runRacer(eps []string, round int) {
 	doRounds(rcs, round)
 }
 
+func runWatcher(eps []string) {
+	// watcher for setting watch on prefixes and also handling the response on watches
+	type watcher struct {
+		prefix string
+		c      *clientv3.Client
+		wc     clientv3.WatchChan
+	}
+
+	prefixes := 36       // total number of prefixes which will be watched upon
+	watchPerPrefix := 10 // number of watchers per prefix
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string)
+	ws := make([]watcher, prefixes*watchPerPrefix)
+
+	var (
+		wg  sync.WaitGroup
+		pr  *clientv3.PutResponse
+		err error
+	)
+
+	const (
+		chars = "abcdefghijklmnopqrstuvwxyz0123456789" // for generating random keys and setting prefixes
+	)
+
+	// assign watchPerPrefix times watchers on each prefix
+	wg.Add(len(ws))
+	for i := 0; i < len(ws); {
+		for _, prefix := range chars {
+			go func(w *watcher, i int, prefix rune) {
+				defer wg.Done()
+				w.c = randClient(eps)
+				w.prefix = string(prefix)
+				w.wc = w.c.Watch(ctx, w.prefix, clientv3.WithPrefix())
+			}(&ws[i], i, prefix)
+			i++
+		}
+	}
+	wg.Wait()
+
+	done := make(chan bool)
+	// generate and put keys in cluster
+	go func(strlen int, runningTime time.Duration, reqRate int, done chan bool) {
+		client := randClient(eps)
+		defer client.Close()
+
+		rand.Seed(time.Now().UTC().UnixNano())
+		result := make([]byte, strlen)
+		limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
+		ctxt, cancel := context.WithDeadline(ctx, time.Now().Add(runningTime*time.Second))
+		defer cancel()
+		for {
+			// generate key
+			for i := 0; i < strlen; i++ {
+				result[i] = chars[rand.Intn(len(chars))]
+			}
+			key := string(result)
+
+			// limit key put as per reqRate
+			err = limiter.Wait(ctxt)
+			if err != nil {
+				done <- true
+				return
+			}
+
+			prefix := key[0:1]
+			vals := prefixKeysMap[prefix]
+			//maintain a map of key prefix and the values put for the prefix
+			prefixKeysMap[prefix] = append(vals, key)
+
+			pr, err = client.Put(ctxt, key, key)
+			if err == context.DeadlineExceeded {
+				done <- true
+				return
+			}
+
+			if err != nil {
+				log.Printf("Error: %v occured while trying to key: %v, value : %v to kv store.", err, key, key)
+				continue
+			}
+		}
+		done <- true
+	}(10, 30, 20, done)
+	<-done
```

Comment:
ok so then in that case i think we should put some sleep time after we trigger the key generation so that some time is allocated for keys to be put for watchers update before it starts validation. Is that ok?